multiclient/client.go

```diff
@@ -657,46 +657,69 @@ func (mc *Client) SubscribeNewHead(ctx context.Context, ch chan<- *Header) (ethe
 	}), nil
 }
 
-func (mc *Client) subscribeNewHead(ctx context.Context, wg *sync.WaitGroup, url string, ch chan<- *Header) error {
+func (mc *Client) subscribeNewHead(ctx context.Context, wg *sync.WaitGroup, id uint64, ch chan<- *Header) error {
+	logger := log.New("id", id)
 	defer wg.Done()
 
 	for {
-		rc := mc.rpcClientMap.Get(url)
-		if rc == nil {
-			log.Trace("EthClient has been removed", "url", url)
+		rc, existed := mc.rpcClientMap.GetById(id)
+		if !existed {
+			logger.Trace("EthClient has been removed")
 			return nil
 		}
-
-		headerCh := make(chan *types.Header)
-		c := ethclient.NewClient(rc)
-		sub, err := c.SubscribeNewHead(ctx, headerCh)
-		if err != nil {
-			log.Warn("Failed to subscribe new head", "url", url, "err", err)
+		if rc == nil {
+			logger.Warn("The client is not ready, wait for a while")
 		} else {
-		WAIT_NEW_HEADER:
-			for {
-				select {
-				case header := <-headerCh:
-					h := &Header{
-						Client: rc,
-						Header: header,
-					}
-					select {
-					case ch <- h:
-					case <-ctx.Done():
-						return nil
-					}
-				case err := <-sub.Err():
-					log.Warn("Failed during subscription", "url", url, "err", err)
-					break WAIT_NEW_HEADER
-				case <-ctx.Done():
-					return nil
-				}
+			// If we have error, we need to retry
+			err := doSubscribe(ctx, id, rc, ch)
+			if err == nil {
+				return nil
 			}
 		}
+
 		// retry subscribe after retryPeriod
-		time.Sleep(retryPeriod)
-		log.Trace("Retry to subscribe new head", "url", url)
+		t := time.NewTicker(retryPeriod)
```

Comment:
Prefer to use time.Timer.
Declare timer outside for-loop.
```
retryTimer := time.NewTimer(0)
defer retryTimer.Stop()
for {
  ......
  retryTimer.Reset(retryPeriod)
  select {
  case <-retryTimer.C:
  case <-ctx.Done():
    return nil
  }
}
```

Reply:
We cannot use it because we need to stop the timer if it's a successful subscription.